variables.tf

```diff
@@ -4,8 +4,191 @@ variable "name" {
+  type        = bool
+}
+
+variable "instance_type" {
+  description = "What instance type to use by default"
+  type        = string
+}
+
+################################
+# Load Balancers
+################################
+
+variable "target_group_arns" {
+  default     = null
+  description = "A set of aws_alb_target_group ARNs, for use with Application or Network Load Balancing."
+  type        = set(string)
+}
+
+################################
+# Health
+################################
+
+variable "health_check_grace_period" {
+  default     = 300
+  description = "Time in sconds after instance comes into service before checking health"
+  type        = number
+}
+
+variable "health_check_type" {
+  default     = "ELB"
+  description = "(Optional) `EC2` or `ELB`. Controls how health checking is done."
+  type        = string
+}
+
+variable "instance_refresh" {
+  description = "(Optional) If this block is configured, start an Instance Refresh when this Auto Scaling Group is updated. This resource does not wait for the instance refresh to complete."
+
+  type = object({
+    strategy = string
+    triggers = optional(set(string))
+
+    preferences = optional(object({
+      instance_warmup        = optional(number)
+      min_healthy_percentage = optional(number)
+    }))
+  })
+
+  default = {
+    strategy = "Rolling"
+  }
+}
+
+variable "max_instance_lifetime" {
+  default     = null
+  description = "(Optional) The maximum amount of time, in seconds, that an instance can be in service, values must be either equal to 0 or between 604800 and 31536000 seconds."
+  type        = number
+}
+
+variable "wait_for_capacity_timeout" {
+  default     = "10m"
+  description = "(Default: `10m`) A maximum duration that Terraform should wait for ASG instances to be healthy before timing out. Setting this to `0` causes Terraform to skip all Capacity Waiting behavior."
+  type        = string
+}
+
+################################
+# ASG Config
+################################
+
+variable "service_linked_role_arn" {
+  default     = null
+  description = "The ARN of the service-linked role that the ASG will use to call other AWS services"
+  type        = string
+}
+
+variable "suspended_processes" {
+  default     = []
+  description = "A list of processes to suspend for the ASG. The allowed values are `Launch`, `Terminate`, `HealthCheck`, `ReplaceUnhealthy`, `AZRebalance`, `AlarmNotification`, `ScheduledActions`, `AddToLoadBalancer`"
+  type        = list(string)
+}
+
 variable "tags" {
   default     = {}
   description = "User-Defined tags"
   type        = map(string)
 }
+
+################################
+# Mixed Instances
+################################
+
+variable "mixed_instances_policy" {
+  default     = null
+  description = "Object defining how to mix on-demand and spot instances of different types"
+
+  type = object({
+    instances_distribution = optional(object({
```

Comment:
omg when did terraform get an `optional` function. this is great

Reply:
@sblack4 experimental in `0.14` and stable in `0.15`